### PR TITLE
nixos/knot: use module system to combine settings option

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

Before options under settings where not merged by the module system but overwritten which doesn't make sense especially for the list types like remotes or template.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by mweinelt** [comment]
@ofborg build knot.tests

File: nixos/modules/services/networking/knot.nix [hunk 1]
```diff
@@ -226,7 +226,7 @@ in {
       };
 
       settings = mkOption {
-        type = types.attrs;
+        type = types.submodule { freeformType = types.attrs; };
```
[the remark this answers is not in the document]
Comment (vcunat) [reply nixos/modules/services/networking/knot.nix:229]: It's a subset of YAML, basically, specific to knot-dns.

[the remark this answers is not in the document]
Comment (vcunat) [reply nixos/modules/services/networking/knot.nix:229]: I didn't know this stuff really, but at a glance I expect it wouldn't be hard to just copy the YAML definition for this and just replace the conversion to string by this knot-specific thing we have here.

[the remark this answers is not in the document]
Comment (vcunat) [reply nixos/modules/services/networking/knot.nix:229]: I'm no good with NixOS option typing.  [The docs](https://nixos.org/manual/nixos/unstable/#sec-option-types-basic) seem to recommend
```nix
type = types.attrsOf types.anything;
````
or I understand your suggestion as
```nix
type = (pkgs.formats.yaml {}).type;
```

Either way, I don't see sense in attempting much eval-time checking here, as we have detailed checks during build time (running `knotc conf-check`).

[the remark this answers is not in the document]
Comment (infinisil) [reply nixos/modules/services/networking/knot.nix:229]: If it's a subset of YAML, then using `pkgs.formats.yaml` sounds good to me :)

[the remark this answers is not in the document]
Comment (vcunat) [reply nixos/modules/services/networking/knot.nix:229]: Merged ✅

**Pull request merged**